ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,410 @@ let order_f (_, vm_rec) =
+            )
+        )
+        ranked_hosts
+    )
+    grp_ranked_hosts
+
+let log_vms_increasing ~__context vms_increasing =
+  debug "log_vms_increasing:" ;
+  List.iter
+    (fun (vm, size) ->
+      debug "  vm: %s %Ld GB"
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (gib_of_bytes size)
+    )
+    vms_increasing
+
+let log_hosts_increasing ~__context hosts_increasing =
+  debug "log_hosts_increasing:" ;
+  List.iter
+    (fun (h, size) ->
+      debug "  host: %s %Ld GB"
+        (Db.Host.get_name_label ~__context ~self:h)
+        (gib_of_bytes size)
+    )
+    hosts_increasing
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
+  in
+  total_memory_of_vm ~__context policy snapshot
+
+(** Generate 2 layers Map: the outer layer is a Map from `VM_group to ranked_hosts, while each
+    ranked_hosts is a Map from vm_count to `host List, where every host in the list has the
+    same number of running anti-affinity VMs resident on it: vm_count. *)
+let anti_affinity_grp_ranked_hosts ~__context hosts =
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+  |> List.fold_left
+       (fun m group ->
+         VMGroupMap.add group
+           (Xapi_vm_helpers.rank_hosts_by_vm_cnt_in_group ~__context group hosts
+           |> List.fold_left
+                (fun m hs ->
+                  match hs with
+                  | (_, vm_cnt) :: _ ->
+                      IntMap.add vm_cnt (List.map fst hs) m
+                  | [] ->
+                      m
+                )
+                IntMap.empty
+           )
+           m
+       )
+       VMGroupMap.empty
+
+(** Update the 2 layers Map - grp_ranked_hosts when a new VM from the anti-affinity group is
+    planned to run on the host. *)
+let update_anti_affinity_grp_ranked_hosts group host grp_ranked_hosts =
+  let ranked_hosts =
+    try VMGroupMap.find group grp_ranked_hosts
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, group not in it" ;
+      raise Not_found
+  in
+  let vm_cnt_on_host =
+    try
+      IntMap.find_first
+        (fun vm_cnt -> List.mem host (IntMap.find vm_cnt ranked_hosts))
+        ranked_hosts
+      |> fst
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, host not in it" ;
+      raise Not_found
+  in
+  VMGroupMap.update group
+    (fun _ ->
+      Some
+        (ranked_hosts
+        |> IntMap.update vm_cnt_on_host (fun hs ->
+               Option.value ~default:[] hs |> List.filter (fun h -> host <> h)
+               |> fun hs -> match hs with [] -> None | _ as hs -> Some hs
+           )
+        |> IntMap.update (vm_cnt_on_host + 1) (fun hs ->
+               Some (host :: Option.value ~default:[] hs)
```

Comment:
```suggestion
               Option.map (fun hosts -> host::hosts) hs
```

Reply:
When it is None, an one element List: [host] is expected.